### PR TITLE
Add documentation for ragged batching

### PR DESCRIPTION
**Pull request by GuanLuo**

### CODE REVIEW

File: docs/ragged_batching.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  {
+    name: "INPUT"
+    data_type: TYPE_FP32
+    dims: [ -1 ]
+  }
+]
+```
+
+With triton ragged batching, the model will be implemented to expect INPUT shape
+[ -1 ] and an additional batch input, INDEX, shape [ -1 ] which the model will
```
Comment (tanmayv25) [inline docs/ragged_batching.md:104]: the model should use to interpret

File: docs/ragged_batching.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+-->
+
+# Ragged Batching
+
+Triton provides dynamic batching feature, which combines multiple requests for
```
Comment (tanmayv25) [inline docs/ragged_batching.md:31]: Link dynamic batching to the dynamic batching scheduler doc here.

File: docs/ragged_batching.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+the same model execution to provide larger throughput. By default, the requests
+can be dynamically batched only if each input has the same shape across the
```
Comment (tanmayv25) [inline docs/ragged_batching.md:33]: By default, the requests can be dynamically batched only if each input has the same shape across the
requests. In order to exploit dynamic batching for a case where input shapes often varies, the client would need to pad the input tensors in the requests to a fixed shape.

File: docs/ragged_batching.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+requests, which may require client-side padding for the use case where the user
+wants to exploit the dynamic batching feature but the request input often varies
+in shape.
+
+Ragged batching is a feature to avoid explicit padding by allowing user to
+specify which of the inputs doesn't require the shape check. User can specify
+such input (ragged input) by setting `allow_ragged_batch` field in the model
```
Comment (tanmayv25) [inline docs/ragged_batching.md:40]: Include `allow_ragged_batching` in the model configuration doc and then link this to it

File: docs/ragged_batching.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    data_type: TYPE_FP32
+    dims: [ 16 ]
+    allow_ragged_batch: true
+  }
+]
+...
+```
+
+How ragged input are processed in a batch of requests depends on the backend
+implementation. For backends that are known to support ragged batching
```
Comment (tanmayv25) [inline docs/ragged_batching.md:57]: The backends -- such as ONNX Runtime, TensorFlow and TensorRT -- require model to accept inputs as 1-dimensional tensors. These backends concatenate the ragged request inputs into the 1-dimensional tensor.

Also link the ONNX Runtime, TensorFLow and TensorrRT to their respective backend repositories.

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

File: docs/ragged_batching.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+(ONNX Runtime backend, TensorFlow backend, TensorRT backend), they require
```
Comment (deadeyegoodwin) [inline docs/ragged_batching.md:58]: Does pytorch not support ragged batching?

File: docs/ragged_batching.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+Because the concatenated input doesn't track the start and end index for each
+request, the backends also require the model to have additional input(s),
+[batch input](#batch-input), that describe various information about the batch
+formed.
+
+## Batch Input
+
+Batch input is often used in combination with ragged input to provide
+information about the each batch element, such as the element count
```
Comment (deadeyegoodwin) [inline docs/ragged_batching.md:70]: ".. about each..."

File: docs/ragged_batching.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+## Example on Ragged Input and Batch Input
+
+If you have a model that accepts 1 variable length input tensor, INPUT, with
+shape [ -1, -1 ]. The first dimension is the batch dimension, and the second
+dimension is the variable-length content. When the client sends 3 requests of
+shapes [ 1, 3 ], [ 1, 4 ], [ 1, 5 ]. To exploit dynamic batching, the
+straight-forward way to implement this model would expect INPUT shape [ -1, -1 ]
+and assume that all inputs were padded to same length so that all requests
+become shape [ 1, 5 ] and thus Triton can batch and send them to the model.
```
Comment (deadeyegoodwin) [inline docs/ragged_batching.md:88]: "... and send them to the model as a single [ 3, 5 ] tensor."

File: docs/ragged_batching.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+]
+batch_input [
+  {
+    kind: BATCH_ACCUMULATED_ELEMENT_COUNT
+    target_name: "INDEX"
+    data_type: TYPE_FP32
+    source_input: "INPUT"
+  }
+]
+```
```
Comment (deadeyegoodwin) [inline docs/ragged_batching.md:131]: "The above example uses BATCH_ELEMENT_COUNT (protobuf link) type of ragged batching. Other types described in [link] operate similarly."

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: docs/ragged_batching.md (hunk 6 above)
Comment (GuanLuo) [reply docs/ragged_batching.md:58]: No, I don't see it has code to handling ragged batching.

**Review by deadeyegoodwin — APPROVED** [review]

**Review by tanmayv25 — APPROVED** [review]
